oracle/oracle.go

```diff
@@ -49,189 +55,174 @@ type oracleOps struct {
 	if err != nil {
-		return metadata, err
+		return metadata, 0, err
 	}
-	v2req.Header.Add("Authorization", "Bearer Oracle")
-	q := v2req.URL.Query()
-	v2req.URL.RawQuery = q.Encode()
 
-	resp, err := client.Do(v2req)
-	if err != nil {
-		logrus.Warnf("metadata lookup from IMDS v1 endpoint failed")
-		return metadata,
-			fmt.Errorf("error occured while getting instance metadata from Oracle Metadata API. Error:[%v]", err)
-	}
-	if resp.StatusCode != 200 {
-		if resp.StatusCode == 404 {
-			// v2 IMDS endpoint not enabled, try v1
-			v1req, err := http.NewRequest("GET", v1MetadataAPIEndpoint, nil)
-			if err != nil {
-				return metadata, err
-			}
-			q := v2req.URL.Query()
-			v1req.URL.RawQuery = q.Encode()
-
-			resp, err = client.Do(v1req)
-			if err != nil {
-				logrus.Warnf("metadata lookup from IMDS v1 endpoint failed")
-				return metadata,
-					fmt.Errorf("error occured while getting instance metadata from Oracle Metadata API. Error:[%v]", err)
-			}
-			if resp.StatusCode != 200 {
-				return metadata,
-					fmt.Errorf("error querying Oracle metadata service: Code %d returned for url %s", resp.StatusCode, v1req.URL)
-			}
-		} else {
-			return metadata,
-				fmt.Errorf("error querying Oracle metadata service: Code %d returned for url %s", resp.StatusCode, v2req.URL)
-		}
+	for headerKey, headerValue := range headers {
+		req.Header.Add(headerKey, headerValue)
 	}
+	q := req.URL.Query()
+	req.URL.RawQuery = q.Encode()
 
+	resp, err := client.Do(req)
+	if err != nil {
+		return metadata, resp.StatusCode,
+			fmt.Errorf("metadata lookup from [%s] endpoint failed with error:[%v]", endpoint, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return metadata, resp.StatusCode, nil
+	}
 	if resp.Body != nil {
 		defer resp.Body.Close()
 		respBody, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
-			return metadata,
+			return metadata, resp.StatusCode,
 				fmt.Errorf("error while reading Oracle metadata response: [%v]", err)
 		}
 		if len(respBody) == 0 {
-			return metadata,
+			return metadata, resp.StatusCode,
 				fmt.Errorf("error querying Oracle metadata: Empty response")
 		}
 
 		err = json.Unmarshal(respBody, &metadata)
 		if err != nil {
-			return metadata,
+			return metadata, resp.StatusCode,
 				fmt.Errorf("error parsing Oracle metadata: %v", err)
 		}
 	}
+	return metadata, resp.StatusCode, nil
+}
+
+func GetMetadata() (map[string]interface{}, error) {
+	httpHeaders := map[string]string{}
+	httpHeaders["Authorization"] = "Bearer Oracle"
+	var httpStatusCode int
+	var err error
+	var metadata map[string]interface{}
+	metadata, httpStatusCode, err = GetRequest(v2MetadataAPIEndpoint, httpHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	if httpStatusCode == http.StatusNotFound {
+		// Authorization header is safely ignored by v1 endpoint
+		fmt.Printf("Trying v1 endpoint as got %d http response\n", httpStatusCode)
+		metadata, httpStatusCode, err = GetRequest(v1MetadataAPIEndpoint, httpHeaders)
```

Comment:
we don't need the httpHeaders for v1 right? Need to send ""?

Reply:
Even if we pass httpHeaders for v1, it simply ignores them.

Reply:
Passing empty header map now.